src/dune_engine/diff_promotion.ml

```diff
@@ -179,3 +178,27 @@ let promote_files_registered_in_last_run files_to_promote =
   let db = load_db () in
   let db = do_promote db files_to_promote in
   dump_db db
+
+let display_file (file : File.t) =
+  let msg = User_message.Annots.empty in
+  let a = Path.source file.dst in
+  let b = File.correction_file file in
+  Print_diff.print msg a b
+
+let filter_db files_to_promote db =
+  match files_to_promote with
+  | All -> db
+  | These (files, on_missing) ->
+    List.fold_left files ~init:[] ~f:(fun acc file ->
+        match
+          List.find_map db ~f:(fun (f : File.t) ->
+              Option.some_if (Path.Source.equal f.dst file) f)
+        with
+        | Some x -> x :: acc
+        | None ->
+          on_missing file;
+          acc)
+
+let display files_to_promote =
+  load_db () |> filter_db files_to_promote |> List.rev
+  |> Fiber.parallel_iter ~f:display_file
```

Comment:
Doing things sequentially would make things deterministic.

Reply:
Or if you're up for doing a little extra work/refactoring, you could generate all the diffs concurrently but then print them sequentially.

Reply:
that was my initial plan, but one issue is that `Print_diff.print` exits the current fiber. Do you mean we could read the output of the diff tool into a `Pp.t` and print that separately?

Reply:
(diff tools can operate either on 2 files or on 2 directories, but if we copy things into temp dirs the diffs won't display good paths)

Reply:
> Do you mean we could read the output of the diff tool into a Pp.t and print that separately?

Yes, that is what I meant. Also, the function would need to stop raising on failure and return an informative value instead. Which is why it would take a little more refactoring.